### PR TITLE
Accept WCH-Link as DAPv1

### LINKED ISSUE #1395: Probe wch.cn WCH-Link (2a86:8011) support
I gave a try to a cheap [`QYF2 CMSIS DAPLink`][QYF2_DAPLink].

pyOCD can't find the debugger.

```
$ lsusb | grep WCH-Link
Bus 003 Device 068: ID 2a86:8011 wch.cn WCH-Link

$ pyocd list
No available debug probes are connected

$ pyocd --version
0.33.1
```

It looks to be ordinary DAP-CMSIS with exotic VID and PID.
I can successfully access the chip with it using openocd this way:
```
$ openocd -c "adapter driver cmsis-dap; cmsis_dap_vid_pid 0x2a86 0x8011"  -f target/stm32g0x.cfg`
```

[QYF2_DAPLink]: https://www.aliexpress.com/item/1005003059525321.html

### PR DESCRIPTION
**Pull request by eugene-bright**
Closes #1395

### CODE REVIEW

**Comment by eugene-bright** [comment]
Should I also include `VID:PID` of the `WCH-Link` to the `KNOWN_CMSIS_DAP_IDS`?

**Comment by flit** [comment]
Hi @eugene-bright! Please take a look at commit a926e94 that adds support for another non-compliant CMSIS-DAP probe. (Man, I wish people would properly implement the spec… 😩) Those changes match solely against the VID/PID, so you won't need the accepted names change. (It also works on all supported OSes/backends.)

The only change needed should be adding the WCH-Link VID/PID to `KNOWN_CMSIS_DAP_IDS` in `pyocd/probe/pydapaccess/interface/common.py`.

You'll also need to rebase your changes onto the `develop` branch and update this PR to target `develop`.

Finally, please add your copyright below the existing copyrights in the file(s) you edit.

If you're quick, you can get it in before the 0.34 release coming late today or tomorrow! 😉 Thanks!

**Comment by eugene-bright** [comment]
@flit Thank you for the valuable advice. In the related issue I was instructed otherwise (to not try to use matching by `VID:PID`).
I'm worried about the commit a926e94. Will it ever come to `main` from `develop`?

**Comment by flit** [comment]
Lol! 😄 Sorry, I'm behind on issues lately since I've been focusing on preparing for the upcoming release, so I hadn't seen the comments from @elfmimi.

Imo, there's are arguments to be made both for matching by name and ID pair. 

`develop` will be merged to `main` shortly, right before the upcoming v0.34 release is made. Standard development process. It keeps `main` free of potential new bugs and regressions until code on `develop` is ready.

File: pyocd/probe/pydapaccess/interface/pyusb_backend.py [hunk 1]
```diff
@@ -343,7 +351,7 @@ def __call__(self, dev):
 
             # Now read the product name string.
             device_string = dev.product
-            if (device_string is None) or ("CMSIS-DAP" not in device_string):
+            if (device_string is None) or not any(map(lambda s: s in device_string, ACCEPTED_DEVICE_NAMES)):
```
Comment (flit) [inline pyocd/probe/pydapaccess/interface/pyusb_backend.py:354]: Quick Python note: `any()` is the right idea! However, you can simply the inner expression using a comprehension as `any(s in device_string for s in ACCEPTED_DEVICE_NAMES)`. It's faster, too. 😄

**Comment by flit** [comment]
At the minimum, please include the VID/PID in the list in `common.py`.

It's up to you if you want to add a generalised device name matching mechanism like your initial commit for this PR. However, it will need to work with all USB backends for CMSIS-DAP (hidapi, pyusb, pyusb_v2, pywinusb). So if you decide to implement this, please move `ACCEPTED_DEVICE_NAMES` to `common.py` and add a device name matching function there that is imported and used in the backends. (Hrrmm… possibly even a single matching function that takes the device name, vid, and pid…)

**Comment by elfmimi** [comment]
I think I need to leave a comment on that VID(0x2a86) , which my instinct tells me is **not** officially approved.

**Comment by flit** [comment]
@elfmimi Very good point. According to the official [USB-IF vendor IDs list](https://www.usb.org/sites/default/files/vendor_ids032322.pdf_1.pdf), VID 0x2a86 (decimal 10886) is registered to "KITRIS AG". Ummm…

(Aside: Why is the world would the USB-IF list vendor IDs _in decimal_? Have you ever heard of someone refernce a VID in decimal? Of course not. 🤣 )

**Comment by flit** [comment]
Fwiw, Patrick Yang, the CTO of WCH, is [on twitter](https://twitter.com/Patrick_RISCV). Someone might ask him publicly why the WCH USB VID is assigned to Kitris AG. 😜

**Comment by flit** [comment]
There's more to this VID story! Apparently another version (?) of the WCH-Link has the VID 0x1A86. (See #1121) This value is correctly registered to Nanjing Qinheng Microelectronics Co., Ltd., aka WCH-IC.

So either someone made a stupid typo in the code? Or something more nefarious is going on. 🤨 Is this 2a86 variant a clone of the real WCH-Link?

**Comment by eugene-bright** [comment]
Probably `0x2a86` is a clone. When I disassembled it I saw marking on the MCU erased. Probably that's done to make the tracking of the producer harder.

**Comment by eugene-bright** [comment]
The `develop` branch lacks some commits from `main`. That's strange...

File: set_version.sh [hunk 2]
```diff
@@ -0,0 +1,11 @@
+#!/usr/bin/env bash
```
Comment (flit) [inline set_version.sh:1]: Could you please move this script to the `scripts/` directory?

Thanks for adding it! Quite useful.

**Comment by eugene-bright** [comment]
It works for me

```
$ python3.10 -m pyocd list 
  #   Probe/Board       Unique ID      Target  
-----------------------------------------------
  0   wch.cn WCH-Link   0001A0000001   n/a
```

**Review by flit — CHANGES_REQUESTED** [review]
**Requested changes**
Would you mind adding your copyright below existing in modified files?
Please move the version script to `scripts/`.

**Next steps**

👉🏽 Do you mind if I push some additional changes to your branch? (GitHub permissions allow project owners to push to PR source branches, but I always ask first.)

There are actually some further issues I've found in `pyusb_backend.py` and `pyusb_v2_backend.py`. The latter should be checking the device name in addition to interface name (either should be accepted). And the interface name checks for both should use your `is_known_device_string()` function.

Also, use of `is_known_device_string()` needs to be integrated into `hidapi_backend.py` and `pywinusb_backend.py`.

My plan is to go ahead and release v0.34 today while we finish this PR. Then I'll cherry pick the changes over to `main` and release v0.34.1.

**Comment by eugene-bright** [comment]
I'm in Ukraine and have to go to sleep to not die from exhaustion. I'm not interested in having any copyrights (and recommend to Linux kernel `sign-off` practice). Please do whatever you see profitable with the code.
See you tomorrow

**Comment by flit** [comment]
'Night! Take care.

**Comment by eugene-bright** [comment]
I moved the script, but the rest is outside of the scope of this ticket. My adapter doesn't have `bInterfaceClass == 0xff` and thus it's not DAP-Linkv2.
I think of how to unify `pyusb_backend.py` and `pyusb_v2_backend.py`. This files are way too similar. The inheritance could be appropirate in this case.

**Review by flit — APPROVED** [review]

**Comment by flit** [comment]
Right, the pyusb v1 and v2 files are very similar and need to be refactored.

**Pull request merged**